### PR TITLE
Pangolin Resolver

### PR DESCRIPTION
**Pull request by prodesert22**
Protocol: https://app.pangolin.exchange/#/swap
Docs: https://docs.pangolin.exchange/about-pangolin
Adding a resolver for Pangolin connector. The connector PR can be found [here](https://github.com/Instadapp/dsa-connectors/pull/125).
Here are the functions available:

**Pangolin Resolver**:

- `getBuyAmount`: Returns the buy amount of token passing another token for sell
- `getSellAmount`: Returns the sell amount of token passing another token for buy
- `getDepositAmount`: Returns deposit amount of LP
- `getSingleDepositAmount`: Returns single deposit amount of LP
- `getDepositAmountNewPool`: Returns deposit amount of LP in new pool
- `getWithdrawAmounts`: Returns withdraw amounts of LP 
- `getPositionByPair`: Returns details of give address in determined tokenA/tokenB pair 
- `getpooldata`: Returns detail of a give address in determined give LP token address
- `getPosition`:Returns details of a give address in determined give LP tokens addresses

**Pangolin Stake Resolver**:

   - `getLPStakeData`: get lp data of stake in minichefv2 of account
   - `getPIDbyLPAddress`: convert lp address to 
   - `getPNGStakeData`: get PNG stake data of gived staking contract

### CODE REVIEW

File: contracts/protocols/avalanche/pangolin/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+import { DSMath } from "../../../utils/dsmath.sol";
+import { IPangolinRouter, IPangolinFactory, TokenInterface } from "./interfaces.sol";
+import "./library.sol";
+
+contract Helpers is DSMath {
+    /**
+     * @dev get Avalanche address
+     */
+    function getAVAXAddr() public pure returns (address) {
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/pangolin/helpers.sol]: I think we can move this to `PangolinHelpers`

File: contracts/protocols/avalanche/pangolin/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        uint256[] memory amts = router.getAmountsIn(buyAmt, paths);
+        sellAmt = amts[0];
+    }
+
+    function getBuyUnitAmt(
+        TokenInterface buyAddr,
+        uint256 expectedAmt,
+        TokenInterface sellAddr,
+        uint256 sellAmt,
+        uint256 slippage
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/pangolin/helpers.sol]: Let's merge same type of params.
```suggestion
         uint256 expectedAmt,
        uint256 sellAmt,
        uint256 slippage
        TokenInterface buyAddr,
        TokenInterface sellAddr,
```

**Comment by pradyuman-verma** [comment]
Hey @prodesert22, Can you update with the requested changes?

**Comment by prodesert22** [comment]
> Hey @prodesert22, Can you update with the requested changes?

Hey @pradyuman-verma, sorry for the delay, is ready for a review this PR

**Comment by pradyuman-verma** [comment]
> > Hey @prodesert22, Can you update with the requested changes?
> 
> Hey @pradyuman-verma, sorry for the delay, is ready for a review this PR

Alright, Can you attach a screenshot of the testcases?

**Comment by prodesert22** [comment]
> > > Hey @prodesert22, Can you update with the requested changes?
> > 
> > 
> > Hey @pradyuman-verma, sorry for the delay, is ready for a review this PR
> 
> Alright, Can you attach a screenshot of the testcases?

Sorry for delay
**Pangolin Resolver**
image
image

**Pangolin Stake Resolver**
image

**Comment by pradyuman-verma** [comment]
@prodesert22 Thanks, Just one last thing. Can you add <SPDX MIT> license in all solidity files?
Then we can proceed toward deployment.

**Review by pradyuman-verma — APPROVED** [review]

**Comment by prodesert22** [comment]
InstaPangolinResolver: [0x9BC3C8382E1f57Ea443d7dE6Bca9cA7D494c325d](https://snowtrace.io/address/0x9bc3c8382e1f57ea443d7de6bca9ca7d494c325d#code)
InstaPangolinStakeResolver: [0xE3B7B78c87EccBe6d87771F11f0e092aaf5881d6](https://snowtrace.io/address/0xE3B7B78c87EccBe6d87771F11f0e092aaf5881d6#code)

**Pull request merged**